### PR TITLE
Iterable torch formatting

### PR DESCRIPTION
**Pull request by lhoestq**
Used the TorchFormatter to get torch tensors in iterable dataset with format set to "torch".

It uses the data from Arrow if possible, otherwise applies recursive_tensorize.

When set back to format_type=None, cast_to_python_objects is used.

requires https://github.com/huggingface/datasets/pull/5821

close https://github.com/huggingface/datasets/issues/5793

### CODE REVIEW

**Comment by lhoestq** [comment]
@albertvillanova could you take a look at this one ? It directly follows the arrow formatting PR

File: src/datasets/iterable_dataset.py [hunk 1]
```diff
@@ -1118,22 +1148,25 @@ def __init__(
         ex_iterable: _BaseExamplesIterable,
         info: Optional[DatasetInfo] = None,
         split: Optional[NamedSplit] = None,
-        format_type: Optional[str] = None,
+        formatting: Optional[FormattingConfig] = None,
         shuffling: Optional[ShufflingConfig] = None,
         distributed: Optional[DistributedConfig] = None,
         token_per_repo_id: Optional[Dict[str, Union[str, bool, None]]] = None,
+        format_type="deprecated",
```
Comment (albertvillanova) [inline src/datasets/iterable_dataset.py:1155]: Nice you deprecate it. What about in the classes above?
- MappedExamplesIterable
- FilteredExamplesIterable

File: src/datasets/iterable_dataset.py [hunk 2]
```diff
@@ -1118,22 +1148,25 @@ def __init__(
...
     ):
         if distributed and distributed.world_size > 1 and shuffling and shuffling._original_seed is None:
             raise RuntimeError(
                 "The dataset doesn't have a fixed random seed across nodes to shuffle and split the list of dataset shards by node. "
                 "Please pass e.g. `seed=42` in `.shuffle()` to make all the nodes use the same seed. "
             )
+        if format_type != "deprecated":
+            formatting = FormattingConfig(format_type=format_type)
```
Comment (albertvillanova) [inline src/datasets/iterable_dataset.py:1163]: Maybe worth raising a warning?

**Review by albertvillanova — COMMENTED** [review]
Thanks. Some comments below.

**Comment by lhoestq** [comment]
I added tests for the `__array__` case which lets you go from any tensor format to any other tensor format.

I also properly deprecated format_type and added a warning message.

**Review by albertvillanova — APPROVED** [review]
Thank you!!!

**Pull request merged**